connectome_interpreter/utils.py

```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
         df = df.drop_duplicates()
         dfs.append(df)
     return pd.concat(dfs)
+
+
+def make_grid_inputs(v1: int | float | list | set | npt.ArrayLike | pd.Series, v2: int | float | list | set | npt.ArrayLike | pd.Series,
+                     num_layers: int, grid_size: int = 10, timepoints: int | list = 0, device=None) -> Tuple[torch.Tensor, List[Tuple[float, float]]]:
+    """
+    Make a batch of input using combinations of v1 and v2 at different strengths (0 to 1). 
+
+    Args:
+        v1: The first input vector.
+        v2: The second input vector. Its length should match that of `v1`.
+        num_layers (int): The number of layers in the model.
+        grid_size (int, optional): The number of points in the grid. Defaults to 10.
+        timepoints (int|list, optional): The timepoints at which combinations of v1 and v2 are used. Defaults to 0 (the first timepoint).
+        device (str, optional): The device to create the inputs on. If None, if GPU is available, the inputs are created on the GPU. Otherwise CPU.
+
+    Returns:
+        torch.Tensor: A tensor of shape (grid_size**2, len(v1), num_layers).
+        list: A list of grid coordinates.
+    """
+    # first convert to np array
+    v1 = to_nparray(v1, unique=False)
+    v2 = to_nparray(v2, unique=False)
+
+    # error if v1 and v2 have more than 1 dimension
+    if v1.ndim > 1 or v2.ndim > 1:
+        raise ValueError("v1 and v2 should be 1D arrays.")
+
+    # check if length of v1 and v2 are the same
+    if len(v1) != len(v2):
+        raise ValueError("The length of v1 and v2 should be the same.")
+
+    if device is None:
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    sc1 = np.linspace(0, 1, grid_size)
+    sc2 = np.linspace(0, 1, grid_size)
+    sc1_grid, sc2_grid = np.meshgrid(sc1, sc2)
+
+    inputs = np.zeros((grid_size**2, len(v1), num_layers), dtype=np.float32)
+    for i, (s1, s2) in enumerate(zip(sc1_grid.ravel(), sc2_grid.ravel())):
+        inputs[i, :, timepoints] = v1 * s1 + v2 * s2
+
+    # Convert to torch tensor
+    inputs_tensor = torch.from_numpy(inputs).to(device)
+
+    # Create grid coordinates for reference
+    grid_coords = list(zip(sc1_grid.ravel(), sc2_grid.ravel()))
+
+    return inputs_tensor, grid_coords
+
+
+def plot_output_grid(grid_outputs: torch.Tensor | npt.NDArray, grid_coords: List[Tuple[float, float]], selected_index: int | List[int], figsize: tuple = (10, 8),
+                     cmap: str = 'viridis', xlab: str = 'v1 Activation', ylab: str = 'v2 Activation', title: str = None):
```

Comment:
Nit:
```suggestion
def plot_output_grid(
    grid_outputs: torch.Tensor | npt.NDArray,
    grid_coords: List[Tuple[float, float]],
    selected_index: int | List[int],
    *,  # see note
    figsize: tuple = (10, 8),
    cmap: str = 'viridis', 
    xlab: str = 'v1 Activation', 
    ylab: str = 'v2 Activation', 
    title: str | None = None,
):
```
I prefer this style for arglists: it's longer but it makes the code "invariant under refactoring" – it's easier to add or remove arguments or rename the function without having to re-indent everything.

For functions with large numbers of kwargs, especially plotting functions, I like to [use `*` to force arguments to be given as keywords](https://stackoverflow.com/questions/59661042/what-do-single-star-and-slash-do-as-independent-parameters): again, easier to change things if you spell them out.